modules/seqexec/server/src/main/scala/seqexec/server/SeqexecEngine.scala

```diff
@@ -485,6 +473,31 @@ object SeqexecEngine extends SeqexecConfiguration {
     )
   }
 
+  def nextRunnableObservations(qid: QueueId)(st: EngineState): Set[Observation.Id] = {
+
+    def findRunnables(used: Set[Resource], obs: Map[Observation.Id, Set[Resource]], acc: Set[Observation.Id]): Set[Observation.Id] = {
+      if (obs.isEmpty) acc
+      else {
+        val next = obs.find(p => used.intersect(p._2).isEmpty)
+        next match {
+          case None => acc
+          case Some((id, res)) => findRunnables(used ++ res, obs - id, acc + id)
+        }
+      }
+    }
+
+    // Get sequence ids, resources and states
+    val seqInfos = st.sequences.map { case (id, ObserverSequence(_, seq)) => id -> ((seq.resources, st.executionState.sequences.get(id))) }.collect { case (id, (res, Some(s))) => id -> ((res, s.status)) }
+    // Set of resources used by running sequences
+    val used = seqInfos.collect { case (_, (res, status)) if (status.isRunning) => res }.foldRight(Set[Resource]())(_.union(_))
+    // resources for observations in queue that are not yet run
+    val obs = st.queues.get(qid).map(_.queue.map(id => id -> seqInfos.get(id)).collect {
```

Comment:
`queue.map(id => id -> seqInfos.get(id))` can be written `queue.fproduct(seqInfos.get)`